### ISSUE TITLE
WriteZero (failed to write whole buffer) error after sending hundert messages

### ISSUE THREAD

**Issue by ondj** [little-dude/rmp-rpc#5]
Hi, I have a problem with sending messages through RPC to the server. Approximately after 100 messages, I got this error, same on Windows 10 and macOS.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: WriteZero, error: StringError("failed to write whole buffer") }) }', src/libcore/result.rs:860
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic
   5: std::panicking::begin_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: <core::result::Result<T, E>>::unwrap
  10: rmp_rpc::client::Endpoint::process_requests
  11: <rmp_rpc::client::Endpoint as futures::future::Future>::poll
  12: <futures::future::chain::Chain<A, B, C>>::poll
  13: <futures::future::and_then::AndThen<A, B, F> as futures::future::Future>::poll
  14: <futures::future::chain::Chain<A, B, C>>::poll
  15: <futures::future::or_else::OrElse<A, B, F> as futures::future::Future>::poll
  16: <alloc::boxed::Box<F> as futures::future::Future>::poll
  17: <futures::task_impl::Spawn<F>>::poll_future_notify::{{closure}}
  18: <futures::task_impl::Spawn<T>>::enter::{{closure}}
  19: futures::task_impl::std::set
  20: <futures::task_impl::Spawn<T>>::enter
  21: <futures::task_impl::Spawn<F>>::poll_future_notify
  22: tokio_core::reactor::Core::dispatch_task::{{closure}}
  23: <scoped_tls::ScopedKey<T>>::set
  24: tokio_core::reactor::Core::dispatch_task
  25: tokio_core::reactor::Core::dispatch
  26: tokio_core::reactor::Core::poll
  27: tokio_core::reactor::Core::run
  28: hashes::main
  29: __rust_maybe_catch_panic
  30: std::rt::lang_start
  31: main
```

As a server, I used two different Python implementation. 

The source code is simple:

```rust
use std::net::SocketAddr;
use futures::Future;
use rmp_rpc::{Value};
use rmp_rpc::client::Client;
use tokio_core::reactor::Core;

fn main() {
    let mut core = Core::new().unwrap();
    let addr: SocketAddr = "127.0.0.1:6000".parse().unwrap();
    let handle = core.handle();

    let client_future = Client::connect(&addr, &handle)
        .or_else(|e| {
            println!("Connection to server failed: {}", e);
            Err(())
        });

    let client = core.run(client_future).unwrap();

    loop {
        let body = Value::from(vec![
            (Value::from("path"), Value::from("a")),
            (Value::from("size"), Value::from(128)),
            (Value::from("md5"), Value::from("abcd")),
            (Value::from("sha1"), Value::from("abcd")),
            (Value::from("sha256"), Value::from("abcd")),
        ]);

        let future = client.request("send_hash", &[body])
            .and_then(|result| {
                println!("{:?}", result);
                Ok(())
            })
            .or_else(|rpc_err| {
                println!("sub failed: {:?}", rpc_err);
                Err(rpc_err)
            });

        core.run(future).expect("Error");
    }
}
```

I tried Rust 1.19 and 1.21 with the same result.

**Comment by ondj** [comment]
I tried to debug it and the problem is here:
https://github.com/little-dude/rmp-rpc/blob/e57cf135c1a38f9e78b143503a4bbfbfe5718d1e/src/codec.rs#L45-L50

`rmpv::encode::write_value` tried to write data to buffer, but the write fails.

But when I add a new line to start of `encode` function:
```rust
buf.reserve(1024);
```
then it works without problem.

**Comment by little-dude** [comment]
This should be fixed in `0.0.10`

Thanks a lot for reporting _and_ fnding the fix :)

**Comment by ondj** [comment]
Sorry, but I think this is just hotfix and not a correct solution. Because it uses `BytesMut`, which:
> One key difference from Vec<u8> is that most operations do not implicitly grow the buffer. This means that calling my_bytes.put("hello world"); could panic if my_bytes does not have enough capacity.

So now if we want to encode bigger message than 1024 bytes, it panics again. 

I think that only correct solution is first to encode `Value` to `Vec<u8>` and then call [bytes::BytesMut::extend_from_slice](https://carllerche.github.io/bytes/bytes/struct.BytesMut.html#method.extend_from_slice), which guarantees to resize if `buf` has not enough capacity.

**Comment by little-dude** [comment]
Mmh re-reading your comment, I think I should have gone with `extend_from_slice` instead of `reserve` and `put_slice`. It seems equivalent, but `extend_from_slice` is just one line.

**Comment by little-dude** [comment]
Ok I'm using `extend_from_slice` now :)

**Comment by little-dude** [comment]
I released 0.1 with the fix!
Thank you for your patience!
